### ISSUE TITLE
Pass CPU offload flag when loading optimizer state

### ISSUE THREAD

**Issue by mvpatel2000** [pytorch/pytorch#117476]
### 🐛 Describe the bug

Distributed checkpoint loading for optimizers does not use CPU offload flag even when it is properly set via a context manager, e.g. via the `set_optimizer_state_dict` api. 

The following line is the source:
https://github.com/pytorch/pytorch/blob/2200118f59cb55d1ed24782e96330749a51f76e4/torch/distributed/fsdp/_optim_utils.py#L391-L393

Separately, I am not sure why there is an all-gather needed for the optimizer loading. It uses substantially more GPU memory than I would expect, especially compared to model loading. The code is unfortunately a lot more complicated and hard to follow compared to model loading. It also is very slow:
```
[rank0]:[2024-01-15 05:44:31,101] torch.distributed.fsdp._debug_utils: [WARNING] FSDP _flatten_optim_state_dict() profiling:  defaultdict(<class 'float'>, {<Type.RESHARDING: 'resharding'>: 193.71070599550148})
```

### Versions

Torch 2.2

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

**Comment by fegin** [comment]
The behavior is correct. Loading process should not take cpu_offload as the flag. It should load the tensor to GPU where the computation happen. The issue looks like OOM instead of the performance (the bottleneck of resharding seems to be from GPU->CPU copy).

**Comment by wconstab** [comment]
> Separately, I am not sure why there is an all-gather needed for the optimizer loading. It uses substantially more GPU memory than I would expect, especially compared to model loading

@fegin or @wz337  can you also comment on this part?

**Comment by awgu** [comment]
https://github.com/pytorch/pytorch/pull/117261 helps mitigate some of the memory usage when loading optimizer state, but I am not sure if it addresses the full concern.

The reason for all-gathering is due to the mismatch in sharding layout. The saved state follows `DTensor`-based per-parameter sharding, whereas existing FSDP uses flat-parameter sharding, requiring all-gather for the conversion.

**Comment by mvpatel2000** [comment]
Got it, understand need for state dict.

It does not address full concern :(. I am still seeing OOMs for large models

**Comment by fegin** [comment]
Debugging with @mvpatel2000 offline. While https://github.com/pytorch/pytorch/pull/117261 helps, `clone()` is an asynchronous call. So PyTorch won't release the occupied memory until the clone finishes. With a very large model, this can still cause CUDA OOM, a potential solution is to call sync() every to force the memory release. 

I'll draft a PR to address the issue.

**Comment by leng-yue** [comment]
What are the potential reasons for sharding layout mismatch?

**Comment by wz337** [comment]
> What are the potential reasons for sharding layout mismatch?

@leng-yue Is the sharding layout mismatch related to this issue? If not, could you open up a new issue for discussion and provide some minimum repro? Thanks.

**Comment by leng-yue** [comment]
> > What are the potential reasons for sharding layout mismatch?
> 
> @leng-yue Is the sharding layout mismatch related to this issue? If not, could you open up a new issue for discussion and provide some minimum repro? Thanks.

@awgu mentioned loading FSDP weights can cause OOM by loading full weights into GPU. This relates to the second warning in the issue linked to FSDP _flatten_optim_state_dict() profiling. The solution in #117261 and the CPU load flag discussed are both in _shard_orig_param_state, suggesting a connection.

If this isn't related, I can move this to a new issue.

**Comment by chauhang** [comment]
cc: @wz337, @LucasLLC

**Comment by snarayan21** [comment]
Hey, coming back to this issue, which commit(s) addressed the OOM issue? We added a `torch.cuda.synchronize` call to the end of the `_shard_orig_param_state` [function](https://github.com/pytorch/pytorch/blob/d60135e91548783a6cbcfc69ea0fbe4976251b16/torch/distributed/fsdp/_optim_utils.py#L376), which I assume was the solution from debugging from @mvpatel2000 and @fegin, but this is causing NaNs in the loss when resuming from checkpoints using the Adam optimizer.

We'd like to cherry-pick the commits which fixed this issue or bump the torch 2.3.0 nightly version we're using, until torch 2.3.0 is actually released. Thanks!
